lib/rmg/ObjectManager.cpp

```diff
@@ -65,33 +77,64 @@ const rmg::Area & ObjectManager::getVisitableArea() const
 	return objectsVisitableArea;
 }
 
-int3 ObjectManager::findPlaceForObject(const rmg::Area & searchArea, rmg::Object & obj, std::function<float(const int3)> weightFunction, bool optimizer) const
+int3 ObjectManager::findPlaceForObject(const rmg::Area & searchArea, rmg::Object & obj, std::function<float(const int3)> weightFunction, OptimizeType optimizer) const
 {
 	float bestWeight = 0.f;
 	int3 result(-1, -1, -1);
 	
-	for(const auto & tile : searchArea.getTiles())
+	if(optimizer & OptimizeType::DISTANCE)
 	{
-		obj.setPosition(tile);
-		
-		if(!searchArea.contains(obj.getArea()) || !searchArea.overlap(obj.getAccessibleArea()))
-			continue;
-		
-		float weight = weightFunction(tile);
-		if(weight > bestWeight)
+		auto open = tilesByDistance;
```

Comment:
What does this collection contain?

Reply:
It's a copy of queue containing tiles from furthest to nearest. It's precomputed so I store it as class member

Reply:
Is it always precomputed? Any chance to misuse this method with not up-to-date cache?

Reply:
I suggest to introduce some USE_CACHE optimizer flag and if it is not set - we always recalculate cache in this method when we need tilesByDistance. This will at least avoid logical errors in future  when you just change optimization type and it does not work as expected

Reply:
It will completely crash the performance. There is method updateDistances() which is called only when new object is placed on the map. If you don't place objects - no need to recompute distance field.

Reply:
this method findPlaceForObject() is marked as const